### PR TITLE
chore: parallelize mprocs lightning channel opening

### PR DESCRIPTION
**Pull request by tvolk131**
On my local machine (Apple M1 Pro, 32GB RAM) the mprocs startup times before/after are:

Before (master branch):
Attempt 1: 93.3s
Attempt 2: 95.5s
Attempt 3: 94.0s
Attempt 4: 93.8s
Attempt 5: 93.5s
Average: 94.0s

After (PR branch):
Attempt 1: 65.2s
Attempt 2: 64.0s
Attempt 3: 63.7s
Attempt 4: 65.7s
Attempt 5: 64.3s
Average: 64.6s

So this is roughly 30% faster and helps to undo some of the damage caused by merging the LDK gateway

### CODE REVIEW

File: devimint/src/devfed.rs [hunk 1]
```diff
@@ -273,12 +273,14 @@ impl DevJitFed {
...
+                            &[(gw_cln, "CLN"), (gw_lnd, "LND"), (gw_ldk, "LDK")]
+                        } else {
+                            &[(gw_cln, "CLN"), (gw_lnd, "LND")]
+                        };
 
-                    if let Some(gw_ldk) = gw_ldk.get_try().await?.deref() {
-                        open_channel_between_gateways(&bitcoind, gw_ldk, gw_cln).await?;
-                        open_channel_between_gateways(&bitcoind, gw_ldk, gw_lnd).await?;
-                    }
+                    open_channels_between_gateways(&bitcoind, gateways).await?;
```
Comment (dpc) [inline devimint/src/devfed.rs:283]: It would probably be better to parallize N calls to open_channels_between_gateways with a different gateway pair each. Parallizing whole thing, instead of each step in the thing separately allows smoothing out differences in each step times.

File: devimint/src/lib.rs [hunk 2]
```diff
@@ -50,11 +50,11 @@ where
     let (process_mgr, task_group) = cli::setup(args).await?;
     log_binary_versions().await?;
     let dev_fed = devfed::DevJitFed::new(&process_mgr, false)?;
-    let res = cleanup_on_exit(f(dev_fed.clone(), process_mgr.clone()), task_group).await;
     // workaround https://github.com/tokio-rs/tokio/issues/6463
     // by waiting on all jits to complete, we make it less likely
     // that something is not finished yet and will block in `on_block`
     let _ = dev_fed.finalize(&process_mgr).await;
+    let res = cleanup_on_exit(f(dev_fed.clone(), process_mgr.clone()), task_group).await;
```
Comment (dpc) [inline devimint/src/lib.rs:57]: Why is this needed? IIUC this ruins parallelism by a lot.

**Comment by dpc** [comment]
https://github.com/fedimint/fedimint/pull/5826 instead?

**Comment by dpc** [comment]
Well, https://github.com/fedimint/fedimint/pull/5826 doesn't work, so whatevs. ;D

File: devimint/src/lib.rs (hunk 2 above)
Comment (tvolk131) [reply devimint/src/lib.rs:57]: But without this change, we're running the test (`f()`) before waiting for the federation to be finalized. Isn't that a race condition?

To answer your question, the LNv2 tests fail without this change. I suspect this is because channel opens with LDK are slow (which is why mprocs is taking so long to start) and on `master` we open a channel between CLN and LND first, allowing for a quick time-to-first-channel:

https://github.com/fedimint/fedimint/blob/cae1f0c6a75262bcd3645b8ee77cc3cbc0c2fdb5/devimint/src/devfed.rs#L273-L281

Whereas on this branch, the time-to-all-channels is quicker but the time-to-first-channel is slower since we're waiting for all channels to reach a certain step before proceeding to the next.

**Review by elsirion — APPROVED** [review]

**Comment by elsirion** [comment]
Looks good and makes devimint go brrrrr :grin:

**Comment by Kodylow** [comment]
@dpc eric said on the call you might have a concern here, can you post if so otherwise can you land?

**Comment by elsirion** [comment]
Ok, too much flakiness, we shouldn't merge yet :/

**Comment by elsirion** [comment]
Talking to @tvolk131: likely unrelated, just unlucky PR

**Comment by dpc** [comment]
@tvolk131 

> But without this change, we're running the test (f()) before waiting for the federation to be finalized. Isn't that a race condition?

The idea was that particular tests should wait for things it actually needs, so that we don't have to pessimize in ever test.
